repos/system_upgrade/common/actors/checkfstabmountorder/libraries/checkfstabmountorder.py

```diff
@@ -0,0 +1,51 @@
+from leapp import reporting
+from leapp.libraries.stdlib import api
+from leapp.models import StorageInfo
+
+
+def _get_incorrectly_ordered_fstab_entries(fstab):
+    """
+    Retrieve pairs of incorrectly ordered entries in /etc/fstab based on mount point overshadowing.
+
+    :param list[FstabEntry] fstab: fstab entries from StorageInfo
+    :returns: fstab entries with incorrect order
+    """
+
+    for i, fstab_entry in enumerate(fstab):
+        mount_point = fstab_entry.fs_file
+
+        for overshadowing_fstab_entry in fstab[i:]:
+            overshadowing_mount_point = overshadowing_fstab_entry.fs_file
+            if mount_point != overshadowing_mount_point and mount_point.startswith(overshadowing_mount_point):
+                yield fstab_entry, overshadowing_fstab_entry
+
+
+def check_fstab_mount_order():
+    critical_mount_points = ['/var/log', '/var/lib', '/var/lib/leapp']
+
+    storage_info = next(api.consume(StorageInfo), None)
+    if storage_info:
+        overshadow_entries = list(_get_incorrectly_ordered_fstab_entries(storage_info.fstab))
+
+        for overshadowed_entry, overshadowing_entry in overshadow_entries:
+            if overshadowed_entry.fs_file in critical_mount_points:
+                summary = (
+                    'Leapp detected incorrect order of entries in /etc/fstab. '
+                    'The entry with {1} mount point overshadows the entry with {0} mount point.'
+                )
+
+                hint = (
+                    'Reverse the order of entries in /etc/fstab with mount points {1} and {0}.'
+                )
+
+                reporting.create_report([
+                    reporting.Title(
+                        'Detected incorrect order of entries in /etc/fstab, preventing a successful in-place upgrade.'
+                    ),
+                    reporting.Summary(summary.format(overshadowed_entry.fs_file, overshadowing_entry.fs_file)),
+                    reporting.Remediation(hint=hint.format(overshadowed_entry.fs_file, overshadowing_entry.fs_file)),
+                    reporting.RelatedResource('file', '/etc/fstab'),
+                    reporting.Severity(reporting.Severity.HIGH),
+                    reporting.Groups([reporting.Groups.FILESYSTEM]),
+                    reporting.Groups([reporting.Groups.INHIBITOR]),
+                ])
```

Comment:
This could lead to the situation when multiple reports are generated, which we do not want. The expectations it that only one report at maximum will inform about all problems with the order of entries in /etc/fstab. So having e.g.
```
/var/log/leapp
/var/log
/var
/
```
will generate one report, informing user about all entries in incorrect order, with the hint how to fix the order.

Reply:
Changed the behavior, so that there is a single report created with a list of overshadowing mount points in detected order and, in the hint, a possible order without overshadowing. I also added a list of duplicate mount points since these shouldn't be in the /etc/fstab as well.